DEVELOPING.md

```diff
@@ -149,3 +142,39 @@ This should create a build job, and then dispatch it to the build worker.
 You can view the build progress in the web UI or by viewing `/hab/svc/builder-worker/data/876066265100378112/log_pipe-876066265100378112.log`. Replace `876066265100378112` with the group ID output by the `start` command.
 
 Note: you will need to upload additional packages to the core origin for the `core/nginx` build to succeed. Follow the same procedure as for `core/hab-backline`. Currently `core/gcc` and `core/libedit` are required.
+
+### Setting up a package upstream
+
+It is possible to configure the on-premise builder to point to the hosted Builder site as an 'upstream'. This allows new packages from the upstream to get created in the on-premise instance automatically.
+
+In order to do so, create a file called `upstream.toml` with the following content:
+```
+[api]
+features_enabled = "upstream"
+
+[upstream]
+endpoint = "https://bldr.habitat.sh"
+```
+
+Then, issue the following command:
+```
+hab config apply builder-api.default $(date +%s) upstream.toml
+```
+After the config is successfully applied, the services should be configured to use the upstream.
+
+Now, you can test out that the upstream works by trying to install a package that you know exists in the upstream (in the _stable_ channel), but not in the local on-premise builder.
+
+```
+hab pkg install -u http://localhost -z <auth-token> <package>
+```
+
+Initially, you will get a `Package Not Found` error.  Wait for a bit (the package will get synchronized in the background) and try again - this time the install should succeed!
+
+*NOTE*: It is important to understand how the upstream cache is working. Packages that are requested (either via a `hab pkg install`, or even searching or browsing packages in the Web UI) in the local on-premise depot that have newer (or existing) versions in the upstream in the *stable* channel are marked for retrieval in the background. It is only after the background retrieval of the package succeeds that the package then becomes available in the local instance. If there is any failure to retrieve or submit the package, the next retrieval attempt will be triggered only by another request for that package. This functionality is new, and will be refined over time.
+
+Once you have installed a package from upstream, you can load the package into your local depot as follows:
```

Comment:
These following lines are confusing in this context and should probably be removed..